### ISSUE TITLE
[Good First Issue]: Enable parallel execution of TensorFlow Layer 2 python tests

### ISSUE THREAD

**Issue by mryzhov** [openvinotoolkit/openvino#20919]
### Context

Current CI (https://github.com/openvinotoolkit/openvino/blob/c0381ab08d2fcdc51c621196e935aa02c64a4b50/.github/workflows/linux.yml#L869) runs all the python tests sequentially one by one and it takes too much time.
The idea is to run all those tests in parallel by using pytest-xdist plugin, but tests should be modified for that.
Please find bellow the common tests issues:
- Random inputs generation as parametrization arguments. Solution: all random test arguments should be generated inside of the tests 
- Tests create, use and remove the same artifacts (temporary data, models, etc). Hot fix solution: all test data should be generated in the system temporary directory (TMPDIR) which can be unique for each thread. Long term solution: all test data should be generated in the unique temporary directory or with unique names.

### What needs to be done?

- Modify TensorFlow Layer 2 tests to solve issues described in the Context. Tests are located here - https://github.com/openvinotoolkit/openvino/tree/master/tests/layer_tests/tensorflow2_keras_tests
- Enable parallel execution of those test in the CI (https://github.com/openvinotoolkit/openvino/blob/4a0098b26ad6c4cbae97fbe34505fa63c08036eb/.github/workflows/linux.yml#L1057)

### Example Pull Requests

https://github.com/openvinotoolkit/openvino/pull/20613

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [Contribution guide](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)

### Contact points

@mryzhov @akashchi

**Comment by LakshmiGayathri19** [comment]
Hello, I am new to contributing to open source projects, and would like to start with this issue. Could you please assign this to me ?
Thanks in advance.

**Comment by LakshmiGayathri19** [comment]
Hey @ilya-lavrenov , Thanks for assigning this to me. I am starting working on this, and will post the updates in the comments.

**Comment by LakshmiGayathri19** [comment]
Hey @ilya-lavrenov , the testcases which I saw in [this](https://github.com/openvinotoolkit/openvino/tree/master/tests/layer_tests/tensorflow2_keras_tests) link are using parametrization arguments with predefined test parameters. For example, if we take [this](https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow2_keras_tests/test_tf2_keras_conv_2d.py) example, [test_data_float32](https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow2_keras_tests/test_tf2_keras_conv_2d.py#L35) variable is pre-defined and is used as the parameter for the test case `test_keras_conv_2d_float32` ([Link](https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/tensorflow2_keras_tests/test_tf2_keras_conv_2d.py#L62)). 

All the test cases which I saw from starting until `test_tf2_keras_conv_2d.py` has the same format of parameter initialization and usage. And in the issue context it is mentioned that `Random inputs generation as parametrization arguments. Solution: all random test arguments should be generated inside of the tests`. However, i don't see any randomly generated parameter inputs for the test cases which I saw until now, all were pre-define "outside the test function" .

So I want to confirm if we should shift the parameter initialization from outside to inside the test function such that when parallelization happen, there wont be any issue. Or are there any random input generations and I should continue looking for them? Just wanted to confirm before I hunt further so that I wont waste my time going down the wrong way.

Thanks!

**Comment by mryzhov** [comment]
Hello @LakshmiGayathri19 , thank you for your interest and help.
You are right, there are no randomly generated parameters in the tests you mentioned. 
Looks like this is another type of issue described in the xdist documentation - https://pytest-xdist.readthedocs.io/en/stable/known-limitations.html. 

I would suggest fixing it by using lists instead of the dictionaries in the test parameters. Could you try it please?

And I would also suggest starting with the second part of the issue and try to run those test in parallel to collect the particular list of problematic tests.

**Comment by LakshmiGayathri19** [comment]
Hello @mryzhov, 

Thank you for clarification. I will try to convert the dict to list in test parameters and also look into running test in parallel.

**Comment by LakshmiGayathri19** [comment]
Hello @mryzhov ,

I made the changes to the `linux.yml` file to enable parallel execution. The line which i changed is:
```python3 -m pytest ${LAYER_TESTS_INSTALL_DIR}/tensorflow2_keras_tests/ --use_new_frontend -m precommit_tf_fe -n auto --junitxml=${INSTALL_DIR}/TEST-tf2_fe.xml```
by adding `-n auto` to enable parallel execution for `TensorFlow 2 Layer Tests - TF FE`. I pushed the changes to my forked repository and I am waiting for the Github workflow to get completed such that I can get the list of test problematic cases. 

But the Github workflow is stuck in `build` stage showing the below status:
`Requested labels: aks-linux-16-cores
Job defined at: LakshmiGayathri19/openvino/.github/workflows/linux.yml@refs/heads/test-parallel-execution
Waiting for a runner to pick up this job...
`.  I am waiting in this since yesterday. I am not understanding why is this taking so long. Did I miss any step or am I going in the right direction ?

Just wanted to confirm with you to be sure that I am doing the right steps.

Thanks

**Comment by mryzhov** [comment]
Hi @LakshmiGayathri19,
aks-linux-16-cores is self-hosted runner and it is not available on the forks, you would need to create PR in openvino-toolokit repository to use it.

**Comment by LakshmiGayathri19** [comment]
Hey @mryzhov ,

I created the PR and the Github workflow ran. Few of the Python unit test cases which belong to `TensorFlow layer 2 - TF FI` failed due to [Known-limitation](https://pytest-xdist.readthedocs.io/en/stable/known-limitations.html). I am working on fixing these test cases.

**Comment by LakshmiGayathri19** [comment]
Hello @mryzhov ,

I recently modified the test cases (only python test cases) which are failing due to enabling parallel execution. And i created a PR for that. But i see that my PR build is failing due to some pip issue. I am not sure if its due to my change. Could you please look into that and let me know if I need to do anything. 


Thanks

**Comment by mryzhov** [comment]
Hi @LakshmiGayathri19, 
There is an error in your changes: https://github.com/openvinotoolkit/openvino/actions/runs/6976462547/job/18988228421#step:26:176
pip error is just a warning.

**Comment by LakshmiGayathri19** [comment]
Thanks for pointing that out @mryzhov . I will fix my test cases. But before raising a PR for these test case changes, is there any way that I can test locally if everything is working fine or not ?

**Comment by mryzhov** [comment]
@LakshmiGayathri19, you can build openvino manually and run those tests.

**Comment by mryzhov** [comment]
Hi @LakshmiGayathri19 , 
Sure, it is not an urgent issue, please keep your time.

**Comment by LakshmiGayathri19** [comment]
Hello @mryzhov,

I tried to build in my local system. But I am getting error in a header file `onnx_openvino_onnx-ml.pb.h` while I am executing the command `cmake --build . --config Release --verbose -j8`. I am following [this](https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/build_windows.md#how-to-build) documentation. 

I am attaching the screenshot of the build error below:

<img width="960" alt="Build-Error" src="https://github.com/openvinotoolkit/openvino/assets/55054579/63dc5f20-c6bf-40b9-91a9-6d20b026f17c">

Do you have any idea how to resolve this error ? Or is it that my system is missing any dependency ?

**Comment by LakshmiGayathri19** [comment]
Hello @mryzhov, I have raised the PR and i see that it is approved by a reviewer. But I see there are some merge conflicts with my branch. So, should I resolve the merge conflicts locally and raise the PR again ?

**Comment by mlukasze** [comment]
@LakshmiGayathri19
yes, please resolve conflicts and push changes.
You may poke Mikhail to check it out and enable CIs if necessary.

**Comment by p-wysocki** [comment]
I am happy to announce that we have created a channel dedicated to Good First Issues support on our [Intel DevHub Discord server](https://discord.gg/wPuqAujS)! Join it to receive support, engage in discussions, ask questions and talk to OpenVINO developers.

**Comment by ilya-lavrenov** [comment]
@rajatkrishna could you please take this one?

**Comment by kamlesh-ops** [comment]
Hi @mryzhov @akashchi I am new to OpenVINO and a novice in open source contribution. Would you like me to take up this issue.
Also, please let me know if some work has been already done( just for the sake of clarification). 
Thanks :)

**Comment by Ruman-12** [comment]
Hey @mryzhov i would like work on this issue.
 could you please assign me this issue?

**Comment by mlukasze** [comment]
@Ruman-12 we've assigned to you another issue, please start there.

**Comment by Ruman-12** [comment]
@mlukasze Thanks

**Comment by kamlesh-ops** [comment]
.take

**Comment by p-wysocki** [comment]
Hello @kamlesh-ops, are you still working on that issue?

**Comment by kamlesh-ops** [comment]
@p-wysocki no sir actually I was busy in some other issue, so I couldn't devote time to this. If you feel you can unassign me, as I may have to take more time in this.

**Comment by Pranshu-S** [comment]
.take

**Comment by Pranshu-S** [comment]
Hi @p-wysocki 
Raised a PR addressing this issue. Do take a look
